### PR TITLE
Hide email address in Sends

### PR DESCRIPTION
**Pull request by eliykat**
Required for PR: https://github.com/bitwarden/web/pull/895

I'm not sure why the tests are failing - it looks like `policy.data` for the policy in the unit tests is filled with garbage data, but I'm not sure how or why.

Should I write new unit tests for the following?
* blocking a user from creating anonymous Send if not permitted
* making sure an email address is not sent in the response model when a Send is viewed and HideEmail is enabled

### CODE REVIEW

File: src/Core/Enums/DisableSendPolicyType.cs
Comment (MGibson1) [inline src/Core/Enums/DisableSendPolicyType.cs:6]: I think this policy item is valid, but I also think it can be misconstrued as applying to access.

We currently can't regulate Send Access at all, but I think we need to be careful how this is explained

File: util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql
Comment (MGibson1) [inline util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql:1]: I believe you need to update the table file in the SQL project as well

**Review by MGibson1 — COMMENTED** [review]
WRT your test questions: There is a fairly significant amount of framework magic going on due to [autofixture](https://github.com/AutoFixture).

I'm happy to meet up with you and walk through everything, but the gist is that it's a dependency injection framework that we've customized to be able to inject semi-curated data (but mostly just random junk) into our tests. It trades initial understanding of tests for simplicity of code, improved organization and reuse, and increased coverage through random input.

The tests you proposed sound like great additions for this work to me

**Comment by MGibson1** [comment]
Also I'm curious as to those license files, what's up there?

File: src/Core/Enums/DisableSendPolicyType.cs
Comment (MGibson1) [reply src/Core/Enums/DisableSendPolicyType.cs:6]: One more thought on this. The Send isn't truly anonymous in the fact that we could trace back who created it. I think for our product, anonymous should refer to true zero knowledge, not just that we don't share it. This is more of a documentation note, I think, but a variable name tweak might be in order too

File: util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql
Comment (addisonbeck) [reply util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql:1]: & the stored procedure files. Pretty much all SQL objects have dedicated files in the SQL project.

**Comment by addisonbeck** [comment]
Is the direction we decided to go with this to totally hide the creator identifier from the access page? The original discussions led me to believe we might just use a different piece of data (like userId) to maintain the "we know who you are" consequences of abuse. 

This is what led to the value being called "CreatorIdentifier" wherever it is already in use, and you may want to use that verbiage instead of "Email" for future proofing.

File: src/Core/Models/Data/DisableSendPolicyData.cs
Comment (cscharf) [inline src/Core/Models/Data/DisableSendPolicyData.cs:5]: Let's rename this to "`SendOptionsPolicyData`"

File: src/Core/Models/Data/DisableSendPolicyData.cs
Comment (cscharf) [inline src/Core/Models/Data/DisableSendPolicyData.cs:7]: Remove this property. Instead just "`public bool DenyHiddenEmail { get; set; }`"

File: util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql
Comment (cscharf) [reply util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql:1]: Ditto

File: src/Core/Services/Implementations/SendService.cs
Comment (cscharf) [inline src/Core/Services/Implementations/SendService.cs:278]: This will need some rework (below)

File: src/Core/Enums/DisableSendPolicyType.cs
Comment (cscharf) [reply src/Core/Enums/DisableSendPolicyType.cs:6]: This type will likely need to go away altogether.

**Review by cscharf — CHANGES_REQUESTED** [review]
Hi @eliykat , here's a few things (some of these are my fault for sure):

* Direction, please create a `new` policy type, "SendOptionsPolicy" and leave the DisableSendPolicy as-is (no migration here, no wonky reverse compatibility issues, etc.
* The new SendOptionsPolicy will, if Send is not disabled overall, allow us to continue to expand options there such as data retention limits, password complexity, force hidden, force password, etc. in the future (not now, but in the future)
* The initial option would be "DenyHideEmail" or something like that, "DisableHiddenEmail" is probably better, you get the gist.

File: src/Core/Enums/DisableSendPolicyType.cs
Comment (eliykat) [reply src/Core/Enums/DisableSendPolicyType.cs:6]: Good point - I'll use HideEmail as the verbiage rather than anonymous (which feels bad to type on qwerty anyway)

**Comment by eliykat** [comment]
Thanks for the feedback all. Summary of further changes across all repos:

* All changes to the `Disable Send` policy have been reverted
* New policy has been added, called the `Send Options` policy
* `DisableHideEmail` is a boolean option within the `Send Options` policy. (I stuck with the `disable` wording but happy to change to `deny` if preferred)
* When `DisableHideEmail` is set:
   * there is a banner at the top of the Send add-edit page to advise the user. This can be expanded if/when further Send options are added that affect this page
   * if the user is creating a new Send, the `hideEmail` checkbox control is disabled
   * if the user is editing an existing Send which does **not** hide their email, the `hideEmail` checkbox control is disabled
   * if the user is editing an existing Send which **does** hide their email (e.g. a Send created before `disableHideEmail` was enabled), the checkbox control is enabled. This gives them the ability to uncheck `hideEmail`, edit and save the Send, effectively making it compliant with the policy. However, if they try to save it without unchecking `hideEmail`, they will get the server error (via toast). If this is too confusing or not intuitive then I can revert it to disabling the Send entirely in this situation.
* SQL tables and stored procedures have been updated to match the migration scripts
* added unit tests for saving Sends with `disableHideEmail = true`. I followed the same pattern as the other tests for `SendService`.

Other notes:
* I kept `hideEmail` rather than `hideCreatorIdentifier` because we are only hiding the email; if we choose to provide an alternative `creatorIdentifier` in its place, then the name still works
* I'd like a unit test for `SendsController.Access` to make sure that the `Send.HideEmail` option is respected; but scaffolding this is beyond me at the moment, I might spend some time later and hit up @MGibson1.

I'll wait for general approval on the above before implementing for other clients, to avoid redoing work if it needs further significant changes.

File: src/Core/Enums/DisableSendPolicyType.cs
Comment (cscharf) [reply src/Core/Enums/DisableSendPolicyType.cs]: `DisableHideCreatorIdentifier` or something to that effect in the property on the policy data itself, yes, this enum would go away all-together however as it is no longer really needed.

File: util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql
Comment (cscharf) [inline util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql:4]: I don't see this default constraint in the Table definition... and isn't the proper way to pre-populate a NOT NULL column.

Please change this (including through line 9) to something like:

```sql
IF COL_LENGTH('[dbo].[Send]', 'HideEmail') IS NULL
BEGIN

    ALTER TABLE [dbo].[Send] ADD [HideEmail] BIT NULL;

END
GO

-- DO OTHER UPDATE STUFF HERE, like Sprocs, views, etc.

-- End of File Below
UPDATE [dbo].[Send] SET [HideEmail] = 0 WHERE [HideEmail] IS NULL;
ALTER TABLE [dbo].[Send] ALTER COLUMN [HideEmail] BIT NOT NULL;
GO
```

Keep in mind as well this kind of `NOT NULL` column change without a default value AND no default value in the SPROC/parameters AND a database deployment that occurs before the server code is fully live/published will cause creating/editing Sends to break momentarily. Not necessarily the end of the world, but something to keep in mind (I don't think we use default values anywhere else either for these `NOT NULL` columns, but personally I'm not a fan of `BIT NOT NULL` when it doesn't really matter, either it's set explicitly or it's not, the fallback to the default value of `false` is easy enough to achieve in code using `Nullable<bool>`/`bool?`.

**Review by cscharf — COMMENTED** [review]
Looks great, just a few comments.

File: util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql
Comment (eliykat) [reply util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql]: All good points, this has clarified my thinking on how to handle db changes. I've changed the `HideEmail` column to `NULL` and made it nullable in the models. Given that, I don't think there's any need to populate it with default values, so I've left that part out.

File: src/Core/Services/Implementations/SendService.cs
Comment (eliykat) [reply src/Core/Services/Implementations/SendService.cs]: I'm all about lame nits, so all good

File: util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql
Comment (eliykat) [reply util/Migrator/DbScripts/2021-03-22_00_Send_Add_HideEmail_Column.sql]: A further note on this - I've made `HideEmail` nullable everywhere in the server, because there are multiple sources it could get null values from (e.g. client, database). In the clients, I've checked and fixed null values in the `SendResponseModel` only, because that's the only entry point for a null value. Let me know if I've missed something here.

**Comment by eliykat** [comment]
Some updates from today:
* See comments above
* PRs for desktop and browser are now up, plus some updated UI screenshots in the web PR
* I have changed the checkbox wording to "Hide my email address from recipients." because (a) this makes the "hide" wording consistent and (b) the shorter wording is required to fit in the browser client window, and I thought it should be consistent across all clients.
* if we add other options to the SendOptions policy, we'll need some logic to collate all the strictest options across organizations, similar to the Master Password policy or Password Generator policy. However, it's not required yet with only one option, so I haven't implemented this yet.
* Mobile PR in progress, aim to get it up tomorrow for the end of the week.

File: src/Core/Models/Api/Response/SendResponseModel.cs
Comment (cscharf) [inline src/Core/Models/Api/Response/SendResponseModel.cs:70]: This can be `bool` here in the View Model because you're defaulting the value. It's only the data model that needs to match the table definition (`bool?` = `BIT NULL`)

```suggestion
        public bool HideEmail { get; set; }
```

File: src/Core/Services/Implementations/SendService.cs
Comment (cscharf) [inline src/Core/Services/Implementations/SendService.cs:296]: Please consider: "Due to an Enterprise Policy, you are not allowed to hide your email address from recipients when creating or editing a Send."

OR whatever I had suggested in the clients themselves, don't remember.

**Comment by eliykat** [comment]
I think this should all be looking good now, subject to any feedback from the rest of the team.

**Review by cscharf — APPROVED** [review]

**Pull request merged**
